### DISCUSSION TITLE
Current working directory that follows the active pane (command widget)

### DISCUSSION THREAD

**Discussion by jessevdp** [dj95/zjstatus#107] [Q&A] [answered]
Hi!

I made a little bash script that prints the current working directory in a nice format.

When I execute this script in a pane it correctly outputs the current working directory of that pane. However, in the statusbar the contents of the widget never change.

I'm assuming that the command is run in the context of the Zellij pane that holds the zjstatus status bar. Therefore it doesn't "update" like I would expect... Since the current working dir of the statusbar pane never updates...

Did anyone figure out how to accomplish this?

Thanks!

---

The script:

```bash
#!/bin/bash

if git rev-parse --is-inside-work-tree &>/dev/null; then
    git_root=$(git rev-parse --show-toplevel)
    git_relative_path=$(git rev-parse --show-prefix | sed 's#/$##')
    if [ "$git_relative_path" = '' ]; then
        echo "$(basename "$git_root")"
    else
        echo "$(basename "$git_root")/$git_relative_path"
    fi
else
    if [[ "$PWD" == "$HOME"* ]]; then
        echo "~${PWD#$HOME}"
    else
        echo "$PWD"
    fi
fi
```

How I run it:

```kdl
command_pwd_command "bash -c 'bash ~/.config/zellij/print_current_path'"
```

**Comment by jessevdp** [comment]
Hey @sommerper! Sorry to just mention you like this.... but I noticed in https://github.com/dj95/zjstatus/issues/22 that you played around with some `pwd` related script too. Perhaps you have some insight? Thanks!

**Reply by sommerper** [reply]
Sorry, just got back from the holidays. Did you figure it out?

**Reply by jessevdp** [reply]
No worries! 

I haven’t really figured it out yet, but mainly because how I would want it to work isn’t possible yet. See @dj95’s excellent answer: https://github.com/dj95/zjstatus/discussions/107#discussioncomment-11691037

Or do you have a magic/nice workaround for handling (and updating) `pwd` command output in zjstatus?

**Reply by dj95** [reply]
Some idea, how this could be (at least partly) achieved, might be a shell hook, that writes into a FIFO. The FIFO could then be utilized in a script within the command widget to fetch the latest CWD based on certain conditions.

I've not really tried and and it might impose some bugs within the widget/command handling (I don't really know if events start to fill up the queue in case some command stalls), but this might offer a chance to implement a feature in that way right now.

In case you need some environment variables with details about the state (like the current focused tab, session name or whatever is available in the internal zjstatus state), I'm willing to implement them for you, such that they can be used in such scripts.

**Reply by jessevdp** [reply]
Hey, sorry for the slow reply! I had a very similar thought actually. 

Just spitballing here... It could even be some shell hook that sends some data over a Zellij pipe. That way zjstatus could keep track of the latest communicated CWD for each pane (since Zellij plugins can know what panes are active right?). That probably gives a 90% solution to that dynamic "focussed pane cwd". (90% because there's probably going to be some weird interactions with shell hooks.)

It wouldn't be super clean though.. And with "pane CWD" info coming to Zellij plugins already it might be better to just wait for that? I'd love to have this functionality right now and am happy to hack on it.. But I think waiting for proper support from Zellij might be the best play here.

I've been tracking https://github.com/zellij-org/zellij/pull/3800. It seems like there's recently been some work done on it.

**Reply by dj95** [reply]
No worries! Your idea with sending the CWD sounds quite tempting, but I'd like to try another way first, as sending the CWD via a pipe has several implications (bug-wise and security-wise).

I've now implemented a new `{pipe}` widget, that might simplify the implementation. It displays the last received string from a zellij pipe. This way we could utilize the zellij built in pipes instead of creating new ones. Thus there's only the need in writing shell hooks, that executes the desired command and sending the payloads to the correct zjstatus widget. With this implementation, I try to shift the responsibility of the command to the shell side, as it knows best when the CWD changes. It also has the complete set of environment variables etc and is much more scriptable than commands within zjstatus :D

**Comment by jessevdp** [comment]
Thinking of this some more... all `command` widgets suffer from this! It's less noticeable with a `command_git_branch_command` because you usually use a fresh session per project & thus git repo...

I was really hoping I could get a display similar to my prompt into the status bar so I have al that same info when inside fullscreen TUIs like NeoVim and LazyGit etc.

Because this looks really cool right!?

<img width="1733" alt="Screenshot 2024-12-29 at 09 28 01" src="https://github.com/user-attachments/assets/1f8cf39c-e735-483c-86ab-b082b7549a86" />


---

The one idea I currently have is to hook into Zellij / ZSH and just send all of the metadata to a file or something so the statusbar commands can read from there.. 

Similar to [this setup for changing tab titles](https://jcd.pub/2024/06/24/setting-the-zellij-tab-title-to-the-running-process-in-zsh/) but slightly different & probably slightly more robuust. (I don't think that solution can handle multiple panes in a single tab correctly. You'd need trigger something in zsh in the pane for it to pick up changes..)

(A quick search landed me on another one of your plugins: [zj-hooks](https://github.com/dj95/zj-hooks). Cool!)

I'm afraid this will get pretty gnarly having to handle multiple sessions etc. I'm not sure yet.

Really all I need is info on the `$PWD` of the currently active & focussed Zellij pane of the session.. 🤔 

Perhaps more zj-status users want the same type of functionality? It might make sense to see if there is a way to get `command_NAME_cwd` to be dynamic with a magic config value like `{focussed_pane_cwd}`. (Just dreaming here..)

**Comment by dj95** [comment, marked as answer]
Hi and thanks a lot for sharing all of your thoughts and ideas! I really appreciate your effort.

First of all, I want to elaborate a bit how the plugins work. Plugins are small wasm files, that are loaded into their own pane, which results them in running in a sandbox. E.g. they can only access certain APIs and paths, that are granted by the user (or implicitely by zellij). That means, that wasm plugins only have access to the information received from zellij besides the directory they are started in.

When executing a command (within the command widget), the wasm plugin tells zellij to run the command. So it's neither directly executed by the plugin nor within a shell environment. In some of the last releases, I've implemented a feature to tell zellij the CWD a command is executed in. We could utilize this one in the future to implement the dynamic CWD.

However, there are some challenges. E.g. there can be multiple panes within a tab with different CWDs set. Which one should be used for the commands? One could assume, that it'll always run in the focused pane, but this can also result in errors or other weird behavior, when e.g. the pane contains an elevated shell and the zellij user does not have permissions in this directory.

Another challenge right now is, that the plugins are not capable of retrieving the CWD from panes within the sessions. Thus we need to first wait for an implementation on zellijs side. There's also some issue for this open (which does not really reflect the feature by it's title :D but the title implicates it). Here's the link to the comment: https://github.com/dj95/zjstatus/issues/92#issuecomment-2460538135

I hope this explained the current situation a bit and revealed some insights why it currently is not available.

As soon as we are capable of getting the CWD, I'd happily use your suggestion of the magic `{focused_pane_cwd}` variable for the command widget configuration to accomplish this goal :)

Happy hacking!

---

**Edit** your screenshot looks awesome and clean!

**Reply by jessevdp** [reply]
Ah yes I missed that issue but it’s exactly the same case indeed. I’ll subscribe there. Thanks for getting back to me, and awesome to hear that this is already in the works!

Happy hacking indeed!
